[REEF-1826] REEF_Must_be_Invoked_in_a_Directory_that_Contains_Its_DLLs

rogancarr opened the pull request:
This addressed the issue by

  * Adding the directory to use to source assemblies as a parameter to the `REEF Client`
  * Added a helper function to use the directory of the currently-running executable
  * Updated HelloREEF to use the executable directory instead of the current working directory

JIRA:
REEF-1826

Pull request:
This closes #1826

rogancarr commented:
Appveyor looks like intermittent test failures.

markusweimer commented on lang/cs/Org.Apache.REEF.Client/Common/ClientParametersConfiguration.cs:
```diff
@@ -0,0 +1,34 @@
+
+namespace Org.Apache.REEF.Client.Common
+{
+    public sealed class ClientParametersConfiguration : ConfigurationModuleBuilder
```
Please add XML doc to all `public` types, fields and methods.

markusweimer commented on lang/cs/Org.Apache.REEF.Examples.HelloREEF/HelloREEF.cs:
```diff
@@ -109,7 +109,20 @@ private static IConfiguration GetRuntimeConfiguration(string name)
+                .Build();
+
+            // Combine the two configurations into one
+            var configuration = TangFactory.GetTang()
```
Use `Configurations.Merge()`

rogancarr commented:
Hi @markusweimer ! Thanks for the review!

I like the `set of folders` approach. In the technical implementation, however, I went with a `List` because `Tang`'s `Sets` require that you add items one-by-one to the configuration, but I couldn't find a way to specify that before I know the content of the sets. `Lists` do present the challenge that there could be multiple of the same item specified, but that is dealt with in the `FileSets` implementation.

rogancarr commented:
Good points.

I moved the functionality for specifying global assemblies into `JobRequestBuilder`. In order to maintain the default behavior, I kept the `Inject`able `GlobalAssemblies` parameter, but I moved it to the `JobRequestBuilder`. So you can specify the global assemblies in a config, or manually add them to the `job request`.

markusweimer commented on lang/cs/Org.Apache.REEF.Client/API/JobRequestBuilder.cs:
```diff
@@ -18,25 +18,40 @@
+            // By default, add the current directory
+            if (globalAssemblies.Count == 0)
+            {
+                globalAssemblies.Add(Directory.GetCurrentDirectory());
```
This is problematic. If a user doesn't want the local directory added, how can they achieve that?

How about we change semantics  in `.Submit` to:
```cs
if(globalAssemblies.Count == 0 && NoFilesAddedToJobRequest)
{
  // Add the current working directory & the directory of the current executuable
}
```

rogancarr replied:
I moved this to the `Build()` statement of the `JobRequest`.

However, it must be noted that this *could* break compatibility with existing applications if they add assemblies *and* expect the local directory to be added by default.

markusweimer commented on lang/cs/Org.Apache.REEF.Client/API/JobRequestBuilder.cs:
```diff
@@ -204,5 +244,23 @@ public JobRequestBuilder SetJavaLogLevel(JavaLoggingSetting javaLogLevel)
+        /// </summary>
+        /// <param name="filePath"></param>
+        /// <returns></returns>
+        private static bool IsAssemblyToCopy(string filePath)
```
Where is this used?

rogancarr replied:
`AddGlobalAssemblies` -- this was the previous behavior to only add in JAR, EXE, and DLL from a specified directory.

motus commented on lang/cs/Org.Apache.REEF.Examples.HelloREEF/HelloREEF.cs:
```diff
@@ -109,7 +112,19 @@ private static IConfiguration GetRuntimeConfiguration(string name)
+            var configuration = Configurations.Merge(clientConfiguration, runtimeConfiguration);
+
+            // Execute the HelloREEF, with these parameters injected
+            TangFactory.GetTang().NewInjector(configuration).GetInstance<HelloREEF>().Run();
```
No need for a joint `configuration` variable: you can simply write
```C#
TangFactory.GetTang()
    .NewInjector(clientConfiguration, runtimeConfiguration)
    .GetInstance<HelloREEF>()
    .Run();
```

rogancarr commented:
Small fixup needed!
-------

Thanks @markusweimer  and @motus ! I updated per your suggestions.

One important thing, as I pointed out above, is that this could now potentially cause a breaking change if there are clients that are adding the current working directory _and_ adding assemblies to the job with the `JobRequestBuilder.AddGlobalAssembly()` method.

rogancarr commented on lang/cs/Org.Apache.REEF.Client/API/JobRequestBuilder.cs:
```diff
@@ -18,32 +18,53 @@
         public JobRequest Build()
         {
+            // If no global assemblies were added, optionally default to those in the working directory
+            if (!_globalAssembliesAdded && _defaultToGlobalAssembliesInWorkingDirectory && _globalAssemblies.Count == 0)
```
This is now the logic for adding the current working directory:
- No assemblies have been added
- The parameter `DefaultToGlobalAssembliesInWorkingDirectory` has been set to true (`Tang` default)
- The injected global assemblies list is empty

rogancarr commented on lang/cs/Org.Apache.REEF.Client/API/JobRequestBuilderFactory.cs:
```diff
@@ -27,13 +27,12 @@ namespace Org.Apache.REEF.Client.API
         [Inject]
-        private JobRequestBuilderFactory(
-            [Parameter(typeof(DriverConfigurationProviders))] ISet<IConfigurationProvider> configurationProviders)
+        private JobRequestBuilderFactory(IInjector injector)
```
Because as written we would need to inject everything into `JobRequestBuilderFactory` that we want to inject into `JobRequestBuilder`.

I added a new parameter to be injected in `JobRequestBuilder`, and rather than propagate that request to the `Factory`, I simply have the `Factory` ask for a handle on the Injector.

rogancarr commented:
I like this approach:

So going forward, global assemblies have to be added using the `JobRequestBuilder` APIs. If no assemblies were added, then we use the current working directory as before.

markusweimer commented:
> So going forward, global assemblies have to be added using the JobRequestBuilder APIs. If no assemblies were added, then we use the current working directory as before.

+1 Also, we might want to log a `WARNING` when we do that, as jobs with no DLLs are fundamentally malformed.

markusweimer commented:
@rogancarr The code was merged, but referenced the wrong PR in the commit. Can you please close this PR?

rogancarr commented:
Closing
